### PR TITLE
Browse Diff: Delete and edit files for artificial commits

### PR DESCRIPTION
**Pull request by gerhardol**
Similar to the Delete/Edit options for Commit popup
Note: All resulting changes are unstaged

Part of #4031

Changes proposed in this pull request:
 - Menu option
 
Screenshots before and after (if PR changes UI):
- Screenshot from slightly different version, Stage is in another PR
image

How did I test this code:
 - Manually

Has been tested on (remove any that don't apply):
 - GIT 2.14.2
 - Windows 7 and Win10

### CODE REVIEW

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -3088,6 +3105,16 @@ private string GetMonoVersion()
             return null;
         }
 
+        private bool DeleteSelectedFiles()
+        {
+            if (DiffFiles.Focused)
+            {
+                diffDeleteFileToolStripMenuItemClick(this, null);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormBrowse.cs]: please don't call handlers directly, extract the share functionality into own method

**Comment by gerhardol** [comment]
FormBrowse is much too big, I agree.
If you will break out DiffFiles, consider sharing that with FormCommit, or make three classes that share code. These additions origins in FormCommit (with the intention of replacing that though).
The following change is a reason to separate the forms:
https://github.com/gitextensions/gitextensions/pull/4084/files#diff-6d931ecb996aeb8676180da98e3f6861R2316

This issue is related to #4092 #4087 
If #4087 is merged but this is not, the tool strip separator should be added separately.

I am cool with waiting but I have an argument to merge in that it does not make the problem more complex, just slightly bigger. The "total" solution may be easier to see than bits and pieces (but impossible to review!)

File: GitUI/CommandsDialogs/RevisionDiff.cs [hunk 2]
```diff
@@ -293,6 +297,14 @@ private void DiffContextMenu_Opening(object sender, CancelEventArgs e)
...
             resetFileToToolStripMenuItem.Enabled = !isCombinedDiff;
 
+            this.diffEditFileToolStripMenuItem.Visible =
+               this.diffDeleteFileToolStripMenuItem.Visible =
+               isExactlyOneItemSelected && !DiffFiles.SelectedItem.IsSubmodule && selectedRevisions[0].IsArtificial();
+
+            this.diffToolStripSeparator13.Visible = isExactlyOneItemSelected &&
+                (!DiffFiles.SelectedItem.IsSubmodule && selectedRevisions[0].IsArtificial() ||
+                DiffFiles.SelectedItem.IsSubmodule && selectedRevisions[0].Guid == GitRevision.UnstagedGuid);
+
```
Comment (RussKie) [inline GitUI/CommandsDialogs/RevisionDiff.cs]: Similar to my https://github.com/gitextensions/gitextensions/pull/4106#pullrequestreview-72685780 lets move the checks out to `RevisionDiffController` so the UI layer is dumb and the logic can be tested.

File: GitUI/CommandsDialogs/RevisionDiff.cs [hunk 3]
```diff
@@ -603,5 +615,78 @@ private void saveAsToolStripMenuItem1_Click(object sender, EventArgs e)
...
+        private bool DeleteSelectedDiffFiles()
+        {
+            if (DiffFiles.Focused)
+            {
+                return DeleteSelectedFiles();
+            }
+            return false;
+        }
+
+        private bool DeleteSelectedFiles()
```
Comment (RussKie) [inline GitUI/CommandsDialogs/RevisionDiff.cs]: Please move this to `RevisionDiffController`
`Module` and any other dependencies should be injected into `RevisionDiffController`'s constructor (similar to `RevisionFileTreeController`)

File: GitUI/CommandsDialogs/RevisionDiff.cs [hunk 4]
```diff
@@ -603,5 +615,78 @@ private void saveAsToolStripMenuItem1_Click(object sender, EventArgs e)
...
+
+        private void diffDeleteFileToolStripMenuItemClick(object sender, EventArgs e)
+        {
+            DeleteSelectedFiles();
+        }
+
+        private void diffEditFileToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            var item = DiffFiles.SelectedItem;
+            var fileName = Path.Combine(Module.WorkingDir, item.Name);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/RevisionDiff.cs]: there are quite a few calls to `Path.Combine(Module.WorkingDir, item.Name)` - that's a good candidate for refactor
perhaps something like `RevisionDiffController.GetFullFilePath(string selectedFile)`

**Comment by RussKie** [comment]
Merging with the same condition as #4106

**Comment by RussKie** [comment]
Please rebase and resolve the conflict and I'll get it merged
Thanks

**Comment by gerhardol** [comment]
After this #4087 will need a merge. I can do refactoring in that, or open a new PR.

**Comment by RussKie** [comment]
A side effect which needs to be fixed - when double clicking on a pending file it opens a diff, which makes no sense

image

**Comment by gerhardol** [comment]
For the double click: I get a FileHistory with irrelevant commits instead. Even worse.
There are likely more issues like that remaining. I am not even sure the hiding/disabling strategy is right or consistent. there are likely several one file vs many selected issues remaining too.

But the most obvious issues should be addressed now.
When the open issues and a few more (at least the major diff commit) is merged, it will be possible to review the result.

@RussKie Really appreciate the effort you put into this!

**Comment by RussKie** [comment]
Can we disable the double click for atrtificial commits?

**Comment by gerhardol** [comment]
I would prefer to stage/unstage at double click, but that differs from normal commits
Otherwise the double click is in DiffFiles_DoubleClick()
We can just igore doubleclick for artificial&&new.
But there are likely several similar issues. I may not even have found all forms yet, #4084 has some improvements though.
But Browse is what most use, that is most important.

**Comment by RussKie** [comment]
I think this may be a confusing UX in the current UI.

**Comment by gerhardol** [comment]
I would prefer to change the default action to stage/unstage for artificial, but the UI should reflect that (bold). (Bold is missing for FileBrowse too).
This is not a new issue, handle it in cleanup for instance with #4098

**Comment by RussKie** [comment]
Let's keep them separate. A bug fix is just that - a fix. An improvement PR should contain a change concerning the improvement. 
This will make it easier to reason about changes, trace issues back to specific commits and, if necessary, roll changes back cleanly.

**Pull request merged**